bindings/experimental/distrdf/python/DistRDF/HeadNode.py

```diff
@@ -306,73 +258,68 @@ def get_inputfiles(self):
-                second list holds the file names of each of the trees in the
-                first list, each tree name can correspond to multiple file
-                names.
+            (tuple): The list of names and aliases of the friends and their
+                corresponding filenames. If the argument to the constructor is
+                not a TTree/TChain the elements of the tuple are None.
         """
-        friend_names = []
-        friend_file_names = []
-
-        # Retrieve the TTree instance
-        tree = self.get_tree()
-
-        # Early return if the dataframe wasn't constructed from a TTree.
-        if tree is None:
-            return FriendInfo()
-
-        # Get a list of ROOT.TFriendElement objects
-        friends = tree.GetListOfFriends()
-        if not friends:
-            # RDataFrame may have been created with a TTree without
-            # friend trees.
-            return FriendInfo()
-
-        for friend in friends:
-            friend_tree = friend.GetTree()  # ROOT.TTree
-            real_name = friend_tree.GetName()  # Treename as string
-
-            # TChain inherits from TTree
-            if isinstance(friend_tree, ROOT.TChain):
-                cur_friend_files = [
-                    # The title of a TFile is the file name
-                    chain_file.GetTitle()
-                    for chain_file
-                    # Get a list of ROOT.TFile objects
-                    in friend_tree.GetListOfFiles()
-                ]
-
-            else:
-                cur_friend_files = [
-                    friend_tree.
-                    GetCurrentFile().  # ROOT.TFile
-                    GetName()  # Filename as string
-                ]
-            friend_file_names.append(cur_friend_files)
-            friend_names.append(real_name)
-
-        return FriendInfo(friend_names, friend_file_names)
+        # Reconstruct friendinfo from the input tree
+        firstarg = self.args[0]
+        if isinstance(firstarg, ROOT.TTree):
+            # RDataFrame(TTree &tree, const ColumnNames_t &defaultBranches = {})
+            # The first argument to the constructor is a TTree or TChain
+            treefriendinfo = ROOT.Internal.TreeUtils.GetFriendInfo(firstarg)
+
+            # Convert to tuples to be able to hash them and cache the clustered ranges
+            friendnamesalias = tuple((str(namealias.first), str(namealias.second))
+                                     for namealias in treefriendinfo.fFriendNames)
+            friendfilenames = tuple(tuple(str(filename) for filename in filenames)
+                                    for filenames in treefriendinfo.fFriendFileNames)
+            friendchainsubnames = tuple(tuple(str(subname) for subname in chainsubnames)
+                                    for chainsubnames in treefriendinfo.fFriendChainSubNames)
+            return friendnamesalias, friendfilenames, friendchainsubnames
+        else:
+            return None, None, None
```

Comment:
Do we need to return three `None` in a tuple or just returning `None` should be enough?

Reply:
This is to make it consistent with the other return type and to have a unique way in which to fill the attributes of the `TreeRange` object (done further down in `TreeHeadNode.build_ranges` that in turn calls `Ranges.get_clustered_ranges`)

Reply:
Mmm but you either get the three or none, right? Perhaps it's because somewhere else this needs to be treated as a single object which is either a tuple or None, instead of three different objects?

Reply:
In the end these three tuples need to be emplaced in the namedtuple `TreeRange` which is defined as
```python
TreeRange = collections.namedtuple(
    "TreeRange", ["start", "end", "treename", "treefilenames", "friendnamesalias", "friendfilenames", "friendchainsubnames", "defaultbranches"])
```

So in particular the attributes `"friendnamesalias", "friendfilenames", "friendchainsubnames"` are already three distinct objects, so we never use the return value of this function as a single tuple

Reply:
Food for thought: wouldn't it be better to actually group those as "friend info" in the TreeRange, instead of making it completely flat? :smile:  Same for start and end, probably.

Reply:
I will incorporate this suggestion in another commit